src/c2ffi/c2ffi.lisp

```diff
@@ -0,0 +1,163 @@
+          (format tmp-include-file-stream "#include \"~A\"~%" input-header-file)
+          (format tmp-include-file-stream "#include \"~A\"~%" tmp-macro-file)
+          :close-stream
+          ;; Invoke c2ffi again to generate the final output.
+          (run-program* *c2ffi-executable* (list* (namestring tmp-include-file) "-o" (namestring output-spec-path)
+                                                  (append arch sys-include-paths))
+                        :output *standard-output*
+                        :ignore-error-status ignore-error-status))))))
+
+(defun spec-path (base-name &key version (arch (local-arch)))
+  (check-type base-name pathname)
+  (make-pathname :defaults base-name
+                 :name (strcat (pathname-name base-name)
+                               (if version
+                                   (strcat "-" version)
+                                   "")
+                               "."
+                               arch)
+                 :type "spec"))
+
+(defun find-local-spec (base-name &key (otherwise nil otherwise?))
+  (let* ((spec-path (spec-path base-name))
+         (probed (probe-file spec-path)))
+    (if probed
+        spec-path
+        (progn
+          (unless otherwise?
+            (error "c2ffi spec file not found for base name ~S" base-name))
+          otherwise))))
+
+(defun ensure-spec-file-exists (header-file-path &key exclude-archs sys-include-paths version)
+  (multiple-value-bind
+        (h-name m-name)
+      (find-local-spec header-file-path :otherwise nil)
+    (if h-name
+        (values h-name m-name)
+        (let ((local-arch (local-arch)))
+          (unless (c2ffi-executable-available?)
+            (error "No spec found for ~S on arch '~A' and c2ffi not found"
+                   header-file-path local-arch))
+          (generate-spec-with-c2ffi header-file-path
+                                    (spec-path header-file-path :arch local-arch :version version)
+                                    :arch local-arch
+                                    :sys-include-paths sys-include-paths)
+          ;; also run c2ffi for other architectures, but tolerate failure
+          (dolist (arch *known-archs*)
+            (unless (or (string= local-arch arch)
+                        (member arch exclude-archs :test #'string=))
+              (unless (generate-spec-with-c2ffi header-file-path
+                                                (spec-path header-file-path :arch arch :version version)
+                                                :arch arch
+                                                :sys-include-paths sys-include-paths
+                                                :ignore-error-status t)
+                (warn "Failed to generate spec for other arch: ~S" arch))))
+          (or (find-local-spec header-file-path)
+              (error "Error finding spec for ~S after running c2ffi" header-file-path))))))
```

Comment:
This error never kicks in because of `find-local-spec`'s _otherwise_ logic, right?